File: TI1/TI1.1 SOAP_JAVA_SINBDD_GR09/02.CLICON/CONUNI_SOAP_JAVA_CLICON_G09/src/main/java/ec/edu/monster/servicios/ClienteConversionSOAP.java
package ec.edu.monster.servicios;

import ec.edu.monster.modelo.Conversion;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.StandardCharsets;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Cliente SOAP para consumir el servicio de conversiones
 * @author ACER NITRO V15
 */
public class ClienteConversionSOAP {
    
    private static final String WS_URL = "http://localhost:8080/CONUNI_SOAP_JAVA_GR09/WSConversion";

    // ========== MÉTODO DE AUTENTICACIÓN ==========

    /**
     * Llama a la operación {@code login} del servidor SOAP para validar credenciales.
     * El cliente NUNCA almacena ni conoce las credenciales correctas.
     *
     * @param usuario   nombre de usuario ingresado
     * @param contrasena contraseña ingresada
     * @return {@code true} si el servidor confirma las credenciales
     * @throws Exception si hay error de comunicación con el servidor
     */
    public boolean login(String usuario, String contrasena) throws Exception {
        String soapRequest =
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                "<soap:Envelope xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\" " +
                                "xmlns:tns=\"http://servicios.monster.edu.ec/\">\n" +
                "    <soap:Body>\n" +
                "        <tns:login>\n" +
                "            <usuario>" + usuario + "</usuario>\n" +
                "            <contrasena>" + contrasena + "</contrasena>\n" +
                "        </tns:login>\n" +
                "    </soap:Body>\n" +
                "</soap:Envelope>";

        String soapResponse = enviarPeticionSOAP(soapRequest);

        // El servidor devuelve <return>true</return> o <return>false</return>
        int startIndex = soapResponse.indexOf("<return>");
        int endIndex   = soapResponse.indexOf("</return>");
        if (startIndex == -1 || endIndex == -1) {
            throw new Exception("Respuesta de autenticación inválida del servidor");
        }
        String resultado = soapResponse.substring(startIndex + 8, endIndex).trim();
        return Boolean.parseBoolean(resultado);
    }


    public Conversion convertirTemperatura(String operacion, double valor) {
        Conversion conversion = new Conversion("Temperatura", operacion, valor, 
            getUnidadOriginal(operacion), getUnidadDestino(operacion));
        
        try {
            double resultado = realizarConversion(operacion, valor);
            conversion.setValorConvertido(resultado);
            conversion.setExitosa(true);
        } catch (Exception e) {
            conversion.setExitosa(false);
            conversion.setMensajeError("Error en conversión de temperatura: " + e.getMessage());
        }
        
        return conversion;
    }
    
    /**
     * Realiza una conversión de longitud
     */
    public Conversion convertirLongitud(String operacion, double valor) {
        Conversion conversion = new Conversion("Longitud", operacion, valor, 
            getUnidadOriginal(operacion), getUnidadDestino(operacion));
        
        try {
            double resultado = realizarConversion(operacion, valor);
            conversion.setValorConvertido(resultado);
            conversion.setExitosa(true);
        } catch (Exception e) {
            conversion.setExitosa(false);
            conversion.setMensajeError("Error en conversión de longitud: " + e.getMessage());
        }
        
        return conversion;
    }
    
    /**
     * Realiza una conversión de peso/masa
     */
    public Conversion convertirPeso(String operacion, double valor) {
        Conversion conversion = new Conversion("Peso/Masa", operacion, valor, 
            getUnidadOriginal(operacion), getUnidadDestino(operacion));
        
        try {
            double resultado = realizarConversion(operacion, valor);
            conversion.setValorConvertido(resultado);
            conversion.setExitosa(true);
        } catch (Exception e) {
            conversion.setExitosa(false);
            conversion.setMensajeError("Error en conversión de peso: " + e.getMessage());
        }
        
        return conversion;
    }
    

    
    /**
     * Realiza la llamada SOAP al servicio web
     */
    private double realizarConversion(String operacion, double valor) throws Exception {
        // Crear el mensaje SOAP
        String soapRequest = crearSOAPRequest(operacion, valor);
        
        // Enviar la petición HTTP
        String soapResponse = enviarPeticionSOAP(soapRequest);
        
        // Parsear la respuesta
        return parsearRespuestaSOAP(soapResponse);
    }
    
    /**
     * Crea el XML SOAP request
     */
    private String crearSOAPRequest(String operacion, double valor) {
        String paramName = getParamName(operacion);
        
        return "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
               "<soap:Envelope xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:tns=\"http://servicios.monster.edu.ec/\">\n" +
               "    <soap:Body>\n" +
               "        <tns:" + operacion + ">\n" +
               "            <" + paramName + ">" + valor + "</" + paramName + ">\n" +
               "        </tns:" + operacion + ">\n" +
               "    </soap:Body>\n" +
               "</soap:Envelope>";
    }
    
    /**
     * Envía la petición SOAP al servicio web
     */
    private String enviarPeticionSOAP(String soapRequest) throws Exception {
        URL url = new URL(WS_URL);
        URLConnection connection = url.openConnection();
        
        connection.setDoOutput(true);
        connection.setRequestProperty("Content-Type", "text/xml; charset=utf-8");
        connection.setRequestProperty("SOAPAction", "");
        
        // Enviar la petición
        try (OutputStream outputStream = connection.getOutputStream()) {
            outputStream.write(soapRequest.getBytes(StandardCharsets.UTF_8));
        }
        
        // Leer la respuesta
        StringBuilder response = new StringBuilder();
        try (InputStream inputStream = connection.getInputStream()) {
            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                response.append(new String(buffer, 0, bytesRead, StandardCharsets.UTF_8));
            }
        }
        
        return response.toString();
    }
    
    /**
     * Parsea la respuesta SOAP y extrae el resultado
     */
    private double parsearRespuestaSOAP(String soapResponse) throws Exception {
        // Buscar el tag <return>
        int startIndex = soapResponse.indexOf("<return>");
        int endIndex = soapResponse.indexOf("</return>");
        
        if (startIndex == -1 || endIndex == -1) {
            throw new Exception("No se encontró el resultado en la respuesta SOAP");
        }
        
        String resultStr = soapResponse.substring(startIndex + 8, endIndex);
        return Double.parseDouble(resultStr.trim());
    }
    
    /**
     * Obtiene el nombre del parámetro para la operación
     */
    private String getParamName(String operacion) {
        switch (operacion) {
            case "celsiusAFahrenheit":
            case "celsiusAKelvin":
                return "celsius";
            case "fahrenheitACelsius":
            case "fahrenheitAKelvin":
                return "fahrenheit";
            case "kelvinACelsius":
            case "kelvinAFahrenheit":
                return "kelvin";
            case "metrosAPies":
            case "metrosAPulgadas":
                return "metros";
            case "piesAMetros":
                return "pies";
            case "pulgadasAMetros":
                return "pulgadas";
            case "kilometrosAMillas":
                return "kilometros";
            case "millasAKilometros":
                return "millas";
            case "kilogramosALibras":
                return "kilogramos";
            case "librasAKilogramos":
                return "libras";
            case "gramosAOnzas":
                return "gramos";
            case "onzasAGramos":
                return "onzas";
            default:
                return "valor";
        }
    }
    
    /**
     * Obtiene la unidad original para la operación
     */
    private String getUnidadOriginal(String operacion) {
        if (operacion.contains("celsiusA")) return "°C";
        if (operacion.contains("fahrenheitA")) return "°F";
        if (operacion.contains("kelvinA")) return "K";
        if (operacion.contains("metrosA")) return "m";
        if (operacion.contains("piesA")) return "ft";
        if (operacion.contains("pulgadasA")) return "in";
        if (operacion.contains("kilometrosA")) return "km";
        if (operacion.contains("millasA")) return "mi";
        if (operacion.contains("kilogramosA")) return "kg";
        if (operacion.contains("librasA")) return "lb";
        if (operacion.contains("gramosA")) return "g";
        if (operacion.contains("onzasA")) return "oz";
        return "unidad";
    }
    
    /**
     * Obtiene la unidad destino para la operación
     */
    private String getUnidadDestino(String operacion) {
        if (operacion.contains("AFahrenheit")) return "°F";
        if (operacion.contains("ACelsius")) return "°C";
        if (operacion.contains("AKelvin")) return "K";
        if (operacion.contains("APies")) return "ft";
        if (operacion.contains("AMetros")) return "m";
        if (operacion.contains("APulgadas")) return "in";
        if (operacion.contains("AMillas")) return "mi";
        if (operacion.contains("AKilometros")) return "km";
        if (operacion.contains("ALibras")) return "lb";
        if (operacion.contains("AKilogramos")) return "kg";
        if (operacion.contains("AOnzas")) return "oz";
        if (operacion.contains("AGramos")) return "g";
        return "unidad";
    }
}

File: TI1/TI1.2 SOAP_DOTNET_SINBDD_GR09/02.CLICON/CONUNI_SOAP_DOTNET_CLICON_G05/CONUNI_SOAP_DOTNET_CLICON_G09/ec/edu/monster/servicios/ClienteConversionSOAP.cs
using System;
using CONUNI_SOAP_DOTNET_CLICON_G09.MiServicio;
using ec.edu.monster.modelo;

namespace ec.edu.monster.servicios
{
    /// <summary>
    /// Cliente SOAP para consumir el servicio de conversiones.
    /// </summary>
    public class ClienteConversionSOAP
    {
        private readonly WSConversionClient clienteSOAP;

        public ClienteConversionSOAP()
        {
            clienteSOAP = new WSConversionClient();
        }

        public bool Login(string usuario, string contrasena)
        {
            return clienteSOAP.login(usuario, contrasena);
        }

        public Conversion ConvertirTemperatura(string operacion, double valor)
        {
            return EjecutarConversion("Temperatura", operacion, valor);
        }

        public Conversion ConvertirLongitud(string operacion, double valor)
        {
            return EjecutarConversion("Longitud", operacion, valor);
        }

        public Conversion ConvertirPeso(string operacion, double valor)
        {
            return EjecutarConversion("Peso/Masa", operacion, valor);
        }

        private Conversion EjecutarConversion(string tipoConversion, string operacion, double valor)
        {
            var conversion = new Conversion(tipoConversion, operacion, valor, ObtenerUnidadOriginal(operacion), ObtenerUnidadDestino(operacion));

            try
            {
                conversion.ValorConvertido = RealizarConversion(operacion, valor);
                conversion.Exitosa = true;
            }
            catch (Exception ex)
            {
                conversion.Exitosa = false;
                conversion.MensajeError = ex.Message;
            }

            return conversion;
        }

        private double RealizarConversion(string operacion, double valor)
        {
            switch (operacion)
            {
                case "celsiusAFahrenheit":
                    return clienteSOAP.celsiusAFahrenheit(valor);
                case "fahrenheitACelsius":
                    return clienteSOAP.fahrenheitACelsius(valor);
                case "celsiusAKelvin":
                    return clienteSOAP.celsiusAKelvin(valor);
                case "kelvinACelsius":
                    return clienteSOAP.kelvinACelsius(valor);
                case "fahrenheitAKelvin":
                    return clienteSOAP.fahrenheitAKelvin(valor);
                case "kelvinAFahrenheit":
                    return clienteSOAP.kelvinAFahrenheit(valor);

                case "metrosAPies":
                    return clienteSOAP.metrosAPies(valor);
                case "piesAMetros":
                    return clienteSOAP.piesAMetros(valor);
                case "metrosAPulgadas":
                    return clienteSOAP.metrosAPulgadas(valor);
                case "pulgadasAMetros":
                    return clienteSOAP.pulgadasAMetros(valor);
                case "kilometrosAMillas":
                    return clienteSOAP.kilometrosAMillas(valor);
                case "millasAKilometros":
                    return clienteSOAP.millasAKilometros(valor);

                case "kilogramosALibras":
                    return clienteSOAP.kilogramosALibras(valor);
                case "librasAKilogramos":
                    return clienteSOAP.librasAKilogramos(valor);
                case "gramosAOnzas":
                    return clienteSOAP.gramosAOnzas(valor);
                case "onzasAGramos":
                    return clienteSOAP.onzasAGramos(valor);

                default:
                    throw new ArgumentException("Operación no soportada: " + operacion);
            }
        }

        private string ObtenerUnidadOriginal(string operacion)
        {
            switch (operacion)
            {
                case "celsiusAFahrenheit":
                case "celsiusAKelvin":
                    return "°C";
                case "fahrenheitACelsius":
                case "fahrenheitAKelvin":
                    return "°F";
                case "kelvinACelsius":
                case "kelvinAFahrenheit":
                    return "K";
                case "metrosAPies":
                case "metrosAPulgadas":
                    return "m";
                case "piesAMetros":
                    return "ft";
                case "pulgadasAMetros":
                    return "in";
                case "kilometrosAMillas":
                    return "km";
                case "millasAKilometros":
                    return "mi";
                case "kilogramosALibras":
                    return "kg";
                case "librasAKilogramos":
                    return "lb";
                case "gramosAOnzas":
                    return "g";
                case "onzasAGramos":
                    return "oz";
                default:
                    return "unidad";
            }
        }

        private string ObtenerUnidadDestino(string operacion)
        {
            switch (operacion)
            {
                case "celsiusAFahrenheit":
                    return "°F";
                case "fahrenheitACelsius":
                    return "°C";
                case "celsiusAKelvin":
                case "fahrenheitAKelvin":
                    return "K";
                case "kelvinACelsius":
                    return "°C";
                case "kelvinAFahrenheit":
                    return "°F";
                case "metrosAPies":
                    return "ft";
                case "piesAMetros":
                case "pulgadasAMetros":
                    return "m";
                case "metrosAPulgadas":
                    return "in";
                case "kilometrosAMillas":
                    return "mi";
                case "millasAKilometros":
                    return "km";
                case "kilogramosALibras":
                    return "lb";
                case "librasAKilogramos":
                    return "kg";
                case "gramosAOnzas":
                    return "oz";
                case "onzasAGramos":
                    return "g";
                default:
                    return "unidad";
            }
        }

        public void Cerrar()
        {
            clienteSOAP?.Close();
        }
    }
}